Reworks the morph. Makes it a more stealth ambush oriented antag

farie82 opened the pull request:
## What Does This PR Do
Makes the morph more stealth and ambush focused. Like a mimic in DnD or such.
1. Morphs now remember forms they learned. They can switch between them on the go (after a short do_after)
2. While morphed morphs are slower. This form is not meant for traveling
3. Morphs are a bit faster while in normal form. Still slower than humans.
4. While morphed morphs can prepare an ambush. This will make their first attack extra powerful and will weaken the target. When somebody touches the morph with their hands they will ambush them as well dealing even more damage. Walking away or attacking things/getting attacked will stop the ambush. Preparing for another 15 seconds will make their disguise perfect. Removing the there is something wrong message when examining. An alert will be shown to notify the user that he's in this mode.
5. Morphs can remember up to 5 forms. After that they will forget the oldest form.
6. Morphs gather food points when eating mobs and lose it when eating items (can't eat items if they have no food points). Bigger mobs mean more points. Humans also have a bonus. They start with a bit so they can use some abilities
7. Morphs can "reproduce" once they ate enough mobs. 5 humans right now or equivalent. A new morph will spawn from the signed up ghost pool. Morphs can only start reproducing on a floor tile. While a ghost is being picked they can't vent crawl.
8. Being eaten by a morph now will turn your suit sensors off if possible. Prisoner jumpsuits are an example of where this is not possible
9. Morphs eating light sources will try to extinguish them to avoid becoming a light source themselves
10. Morphs can no longer turn into certain things. Like effects or things that are way to big
11. Morphs have objectives! EAT AND REPRODUCE!
12. Morphs can open vents/scrubbers at the cost of stored food. Takes a bit to do for the user and then it slowly burns away 4 second total.
13. Eating items as a morph now takes 3 seconds instead of 21
14. Morphs can pass closed airlocks using an ability. This takes 6 seconds and won't work on bolted airlocks.
15. Morphs are now unpushable and unpullable in their true form.
16. New dead sprite for the morph. The old one was... weird
17. The Gluttony ruin now has a matter eater injector instead of the morph injector. It would be problematic to have miners become morphs now since they can reproduce and what not.
18. Turning into an object with the color var set to anything now properly resets your color once you become your true form again as morph

I've also changed the NPC behaviour to correctly use the new mimic spell and ambush ability. They do cheat though

All action button sprites except the pass airlock one were provided by @smokincheez
Dead morph and the pass airlock action button sprites was provided by @Pewtershmitz

Fixes a bug where wirecutters did not return a value when the attack succeeded. Dunno how much this impacted though

Fixes: #13800 since it removes the glut ruin morph injector. Thus no more cheese. Other cases can be dealt with using admins

## Why It's Good For The Game
A morph is currently a bit... boring? It can turn into things but then it has no real ambush ability. If you undo your form you will be to slow to catch them and you do to little damage to actually hurt them enough to slow them down usually.
This PR aims to make morphs a dangerous foe for the crew. And something the crew should be paranoid about when finding a random taser in maintenance.

## Images of changes
![morph_ambush_morph](https://user-images.githubusercontent.com/15887760/104088963-0db6e380-526b-11eb-96d1-ae01e60237c6.gif)
Morph morphing and ambush

![5N7oGaX5ZH](https://user-images.githubusercontent.com/15887760/103302541-371e7680-4a04-11eb-955f-c1782a2c1788.gif)
When somebody tries to use the mimic while in ambush mode

![i1fmMBLgNl](https://user-images.githubusercontent.com/15887760/103373375-0a865f80-4ad5-11eb-964d-297b68034b94.gif)
Open vent ability

![jz0wgV3tx5](https://user-images.githubusercontent.com/15887760/103373467-5507dc00-4ad5-11eb-8129-fd223b70cc60.gif)
Reproduce ability (you spawn on the same tile)

![morph_airlock](https://user-images.githubusercontent.com/15887760/104088967-190a0f00-526b-11eb-8dc6-b14a5817269a.gif)
Morph passing through airlocks

## Changelog
:cl: Farie82 smokincheez Pewtershmitz
add: Added the Mimic spell. Works for morphs. Potential to work for other mobs (doesn't work yet). When used on other objects you will start to remember their form. When used on yourself you will get the option to morph into a remembered form
add: Morphs can now prepare an ambush when morphed. Dealing a ton of damage and will weaken the target.
add: Morphs now gather food points when they eat. This can be spent on eating non mobs and reproducing
add: Morphs can reproduce by eating enough mobs (5 humans will do)
add: Morphs can open welded vents/scrubbers at the cost of stored food
add: Morphs have formal objectives now. EAT AND REPRODUCE
add: Morphs can now pass through closed airlocks which are not bolted using an ability
tweak: Eating items as a morph costs food. 5 per item you eat. Not enough food points means no eating items
tweak: Being eaten by a morph turns your suit sensors off if possible
tweak: Morphs will try to extinguish light giving objects when eating them. Avoiding them to become a light source themselves
tweak: Morphs now use the Mimic spell.
tweak: Morphs are a bit faster when not morphed and slower when morphed.
tweak: Morphs do a bit less damage in normal form
tweak: Morphs can no longer transform into effects or objects to big.
tweak: Eating items as a morph now takes 3 seconds instead of 2
tweak: Morphs are unpushable and unpullable in their true form
tweak: Gluttany's ruin now has a matter eater injector instead of a morph injector
imageadd: Action background for morphs by smokincheez
imageadd: Icon for the mimic spell and the morph ambush, vent open and reproduce ability by smokincheez
imageadd: Alert icon for the morph ambush ability
imageadd: New dead morph sprite by Pewtershmitz
imageadd: Icon for the morph pass through airlock action button sprite by Pewtershmitz
fix: Turning into an object with the color var set to anything now properly resets your color once you become your true form again as morph
/:cl:

Drobe50 commented:
Does this also affect the morphs created by mining loot?

Dashsora commented:
Morphs also desperately need more concise objectives. 

I've seen a midround morph, disguised as an e-dagger, working as a sec offcer's partner, staying hidden until the officer was in danger and then revealing themselves to attack/drag them away... when ahelped, the answer was pretty much that morphs don't have objectives and as such are free to do whatever. This was an extremely frustrating experience and I really, really, think some clearer objectives for them is needed.

farie82 commented:
> Does this also affect the morphs created by mining loot?

Yes any type of morph. Mining morphs are... a pain in the ass for me. I am not sure if they are seen as antags or not



> Morphs also desperately need more concise objectives.
> 
> I've seen a midround morph, disguised as an e-dagger, working as a sec offcer's partner, staying hidden until the officer was in danger and then revealing themselves to attack/drag them away... when ahelped, the answer was pretty much that morphs don't have objectives and as such are free to do whatever. This was an extremely frustrating experience and I really, really, think some clearer objectives for them is needed.

Aye I hate those morphs as well. I'll think about adding some goals to make them unable to be "friendly" (disgusting).
I hope that the ability to reproduce and their abilities relying on eating mobs also helps with that

farie82 commented:
New update:
Morphs can open vents now by spending collected food. Gif coming soon.
Morphs have actual objectives now. All be it simple.
Morphs no longer can become effects and big things like the DNA vault or the dead clockwork god on lavaland. (might still not have covered all the edge cases though)
They start with a bit of stored food now

Sprites will be added soon. Made again by smokincheez!

procdrone commented:
Greetings mimic. 

Good change.

Spacemanspark commented:
Very well made. Looking forward to this.

farie82 commented:
TM was done. Quite some issues were found and I got a lot of suggestions. Here a summary:

- Examine was broken... making the morphs appear normal while disguised. Fixed
- Morphs could split in vents. Fixed now by making them unable to do that + they can't vent crawl while the ghost player gets picked.
- I've nerfed their base speed and morphed speed a bit
- Their healing is now based on the food gained from the being eaten. It is a 1 on 1 conversion. So 30 healing from a human. Was 50 for all
- Fixed a bug where morphs could hide under vents when they disguised themselves as hidden maint drones or other hidden beings. They will now always be positioned on the Mob layer
- When using the ambush ability you now won't get the food left message
- Fixed a rogue pixel on the dead morph sprite (maybe another sprite will be made. Maybe not)
- Morphs can now pass through closed airlocks using an ability. Bolted airlocks will prevent the morph from passing it. No more having to break down maint doors when you spawn
- Fixed a bug where the ambush do_after didn't stop when unmorphing
- Fixed a runtime when you morphed twice without unmoprhing
- Fixed a runtime when remembering more than the limit of the forms
- Morphs can now not be pulled/pushed when they are in their true form. Morphs are slower now so getting away is painful already. Cheesing them getting away by pushing/pulling them is not possible now when they are unmorphed.

TODOs:
Add a sprite for the pass through doors ability.
Maybe add another sprite for the dead morph

SteelSlayer commented on code/game/gamemodes/miniantags/morph/spells/morph_spell.dm:
```diff
@@ -0,0 +1,36 @@
+		return FALSE
+	return ..()
+
+/obj/effect/proc_holder/spell/morph/before_cast(list/targets, mob/living/simple_animal/hostile/morph/user)
+	user.use_food(hunger_cost)
+	if(hunger_cost)
+		to_chat(user, "<span class='boldnotice'>You have [user.gathered_food] left to use.</span>")
```
The way this is set up makes abilities which cost food give confusing feedback in chat: 

![image](https://user-images.githubusercontent.com/42044220/103852766-30ca7300-5072-11eb-9edc-3ee107148cee.png)

`user.use_food(hunger_cost)` should probably go further down in the chain of spell procs or something.

farie82 replied:
Aye I saw that as well while testing. Issue with fixing this is that I have to refactor the spell code even more :(
The costs are currently deducted before a cast. So if the user cancels the cast midway they will get a refund then the message will make more sense. Currently the `after_cast` proc is also always called no matter what `cast` returns so putting it there will lead to even worse messages. You have 30 left. You have 20 left (since it gets reversed before the `after_cast` gets called.

I think I'm gonna leave this one in till that refactor

SteelSlayer replied:
Ah alright, I don't blame you, spell code is confusing/annoying to work with. Refactor is out of scope.

SteelSlayer commented on code/game/gamemodes/miniantags/morph/spells/pass_airlock.dm:
```diff
@@ -0,0 +1,50 @@
+// TODO refactor when spell code is component based instead of OO based
```
Component based? :eyes:

farie82 replied:
Not the components as in DM but more like described here https://en.wikipedia.org/wiki/Component-based_software_engineering 
The targeting on it's own can be a datum or so which you can just put on your spell at will. Which will reduce the OO madness where you can't make a morph AOE spell and a morph click based spell
Soon:tm:

SteelSlayer replied:
Oh ok I was thinking datum components, and I got excited for a minute :smile: Still, sounds like a good idea.

SteelSlayer commented on code/game/gamemodes/miniantags/morph/spells/pass_airlock.dm:
```diff
@@ -0,0 +1,50 @@
+	name = "Pass Airlock"
+	desc = "Reform yourself so you can fit through a non bolted airlock. Takes a while to do and can only be used in a non disguised form."
+	clothes_req = FALSE
+	charge_max = 100
```
```suggestion
	charge_max = 10 SECONDS
```

SteelSlayer commented on code/game/gamemodes/miniantags/morph/spells/reproduce.dm:
```diff
@@ -0,0 +1,42 @@
+	var/mob/C = pick(candidates)
+
+	var/mob/living/simple_animal/hostile/morph/new_morph = new /mob/living/simple_animal/hostile/morph(get_turf(user))
+	var/datum/mind/player_mind = new /datum/mind(C.key)
```
I don't think I've ever seen a new mind created like this when used for ghost roles. The observer should already have a mind (hopefully). Is there a reason you did this?

farie82 replied:
If you transfer the mind from the observer to the morph then they will keep their old objectives and memories. Even their martial arts and the like...
I don't want a hijack morph with carp :P

SteelSlayer replied:
That's true I suppose.. which means it already happens with human based ghost roles, abductors, SITs and so on. :flushed:

farie82 commented on code/datums/spells/mimic.dm:
```diff
@@ -0,0 +1,198 @@
+	RegisterSignal(user, COMSIG_PARENT_EXAMINE, .proc/examine_override)
+	RegisterSignal(user, COMSIG_MOB_DEATH, .proc/on_death)
+
+	SEND_SIGNAL(user, COMSIG_MAGIC_MIMIC_CHANGE_FORM, form)
```
The spell here is not only for morphs. I can however override the specific instance to avoid the signal. Which does sound better yeah

AllStickyGithub commented:
> mining morphs are... a pain in the ass for me. I am not sure if they are seen as antags or not
If we are going to make a powerful antag out of morphs I am not sure we can justify leaving a free antag item on lavaland. The lavaland morph was never intended as an antagonist though. It was more like something miners could use to goof around with and miners have been punished by admins for abusing it. The real problem I see is the reproduction. If a miner morphs starts to reproduce what rules should the offspring follow?

I loved the TM though and if the gluttony ruin has to go for this it will still be worth it.

farie82 commented:
Last two sprites added which were provided by @Pewtershmitz
PR should be fully done now :tm:

Spacemanspark commented:
I'd be fine with removing the mining gluttony ruin

TheYeetster1 commented:
Was playing a round, Gluttony's Ruin Morph disguised as the bar sign, which is 2 tiles wide. I think you asked to be notified if you missed anything that they shouldn't be able to be disguised as? @farie82

farie82 commented:
> Was playing a round, Gluttony's Ruin Morph disguised as the bar sign, which is 2 tiles wide. I think you asked to be notified if you missed anything that they shouldn't be able to be disguised as? @farie82

Yep! Thanks for the comment. I'll add it to the blacklist. I'm sure I won't get all the items but getting most "common" ones on the station would be good

Necaladun reviewed (approved):
Looks way more fun to me

Spacemanspark commented:
@farie82 A morph was also able to disguise itself as sparks recently. Not sure if that's fixed in this PR but worth mentioning.

farie82 commented:
> @farie82 A morph was also able to disguise itself as sparks recently. Not sure if that's fixed in this PR but worth mentioning.

Yeah that is fixed here. Ive made it impossible to become an effect

farie82 commented:
Small update, the gluttony ruin now has a matter eater implanter instead of the morph injector. Having miners become the new type of morph will be problematic to say the least

dearmochi commented on code/game/gamemodes/miniantags/morph/spells/reproduce.dm:
```diff
@@ -0,0 +1,42 @@
+		return FALSE
+
+/obj/effect/proc_holder/spell/morph/reproduce/cast(list/targets, mob/living/simple_animal/hostile/morph/user)
+	to_chat(user, "<span class='sinister'>You prepare to split in two. Making you unable to vent crawl!</span>")
```
```suggestion
	to_chat(user, "<span class='sinister'>You prepare to split in two, making you unable to vent crawl!</span>")
```

farie82 commented on code/datums/spells/mimic.dm:
```diff
@@ -0,0 +1,206 @@
+		user.pixel_x = initial(user.pixel_x)
+		user.layer = MOB_LAYER // Avoids weirdness when mimicing something below the vent layer
+
+	playsound(user, "bonebreak", 75, TRUE)
```
**CRUNCH**

S34NW commented:
Is this in a ready state (minus the conflict?)

farie82 commented:
> Is this in a ready state (minus the conflict?)

Yep yep. All finishes
